gluten-core/src/main/scala/org/apache/spark/sql/execution/datasources/GlutenColumnarRules.scala

```diff
@@ -101,67 +104,50 @@ private case class ColumnarToFakeRowLogicAdaptor(child: LogicalPlan)
+ * now.
  */
-case class ColumnarToFakeRowAdaptor(child: SparkPlan) extends ColumnarToRowTransition {
+case class FakeRowAdaptor(child: SparkPlan) extends ColumnarToRowTransition {
   if (child.logicalLink.isDefined) {
     setLogicalLink(ColumnarToFakeRowLogicAdaptor(child.logicalLink.get))
   }
 
   override def output: Seq[Attribute] = child.output
 
   override protected def doExecute(): RDD[InternalRow] = {
-    child.executeColumnar().map { cb => new FakeRow(cb) }
+    if (child.supportsColumnar) {
+      child.executeColumnar().map(cb => new FakeRow(cb))
+    } else {
+      val r2c = BackendsApiManager.getSparkPlanExecApiInstance.genRowToColumnarExec(child)
+      r2c.executeColumnar().map(cb => new FakeRow(cb))
+    }
   }
 
   // For spark 3.2.
-  protected def withNewChildInternal(newChild: SparkPlan): ColumnarToFakeRowAdaptor =
+  protected def withNewChildInternal(newChild: SparkPlan): FakeRowAdaptor =
     copy(child = newChild)
 }
 
-case class LoadArrowData(child: SparkPlan) extends UnaryExecNode {
-  override protected def doExecute(): RDD[InternalRow] = {
-    throw new UnsupportedOperationException(
-      "LoadArrowData does not support the execute() code path.")
-  }
-
-  override def nodeName: String = "LoadArrowData"
-
-  override def supportsColumnar: Boolean = true
+object GlutenColumnarRules {
 
+  // TODO: support Insert clause
 
-  override protected def doExecuteColumnar(): RDD[ColumnarBatch] = {
-    child.executeColumnar().mapPartitions { itr =>
-      BackendsApiManager.getIteratorApiInstance.genCloseableColumnBatchIterator(
-        itr.map { cb =>
-          ArrowColumnarBatches.ensureLoaded(ArrowBufferAllocators.contextInstance(), cb)
-        })
+  def isGlutenInsertInto(cmd: DataWritingCommand): Boolean = {
+    cmd match {
+      case command: InsertIntoHadoopFsRelationCommand =>
+        command.fileFormat.isInstanceOf[GlutenParquetFileFormat]
+      case _ => false
     }
   }
 
-  override def output: Seq[Attribute] = {
-    child.output
-  }
-
-  override protected def withNewChildInternal(newChild: SparkPlan): SparkPlan =
-    copy(child = newChild)
-}
-
-object GlutenColumnarRules {
-  // Load when fallback to vanilla columnar to row
-  // Remove it when supports all the spark type velox columnar to row
-  case class LoadBeforeColumnarToRow() extends Rule[SparkPlan] {
-    override def apply(plan: SparkPlan): SparkPlan = plan.transformUp {
-      case c2r @ ColumnarToRowExec(_: ColumnarShuffleExchangeExec) =>
-        c2r // AdaptiveSparkPlanExec.scala:536
-      case c2r @ ColumnarToRowExec(_: ColumnarBroadcastExchangeExec) =>
-        c2r // AdaptiveSparkPlanExec.scala:546
-      case ColumnarToRowExec(child) => ColumnarToRowExec(LoadArrowData(child))
+  case class NativeWritePostRule(session: SparkSession) extends Rule[SparkPlan] {
+    override def apply(p: SparkPlan): SparkPlan = p match {
+      case rc @ DataWritingCommandExec(cmd, child) if isGlutenInsertInto(cmd) =>
```

Comment:
If child is a row-based operator, adding a FakeRowAdaptor here will report an error, right? Our current code does not remove the ColumnarToRow in front of AQE.
![image](https://github.com/oap-project/gluten/assets/11972570/f6e7c65f-c105-4227-87eb-6902eef9fcb6)